component/mimir/rules/diff.go

```diff
@@ -0,0 +1,117 @@
+package rules
+
+import (
+	"bytes"
+
+	mimirClient "github.com/grafana/agent/pkg/mimir/client"
+
+	"gopkg.in/yaml.v3"
+)
+
+type RuleGroupDiffKind string
+
+const (
+	RuleGroupDiffKindAdd    RuleGroupDiffKind = "add"
+	RuleGroupDiffKindRemove RuleGroupDiffKind = "remove"
+	RuleGroupDiffKindUpdate RuleGroupDiffKind = "update"
+)
+
+type RuleGroupDiff struct {
+	Kind    RuleGroupDiffKind
+	Actual  mimirClient.RuleGroup
+	Desired mimirClient.RuleGroup
+}
```

Comment:
I've been trying to encourage people to keep the publicly exposed API of new components small to have greater flexibility over how our code gets used. It doesn't look like this needs to be exported; I'd recommend unexporting it. D

itto with everything in `events.go`.